### DISCUSSION TITLE
Hydra performance issues when used with cockroachdb

### DISCUSSION THREAD

**Discussion by daastek** [ory/hydra#2499] [Q&A] [answered]
We are doing some performance testing of hydra with cockroach db. We are running hydra in k8s and can scale as much as needed. Cockroachdb is also running on the same K8s cluster with 3 vCPU and 12GB memory. When we run a load test with 150 requests per second for the client_credentials flow, we are getting response time of 2-4 seconds. The cockroachdb is not showing any considerable load while the test is run.(only 6% cpu use and 19ms query lag).  When the same test is run against a mysql cluster of similar size, the response time is less than 40ms. I have attached some traces from jaeger during the test.

Is there any known performance issues with hydra when configured with cockroachdb?
![image](https://user-images.githubusercontent.com/67271546/116038136-556e4900-a669-11eb-8e12-85fd6a77cb00.png)

**Comment by aeneasr** [comment, marked as answer]
Probably bcrypt: https://www.ory.sh/hydra/docs/benchmark/#bcrypt https://www.ory.sh/hydra/docs/security-architecture/#bcrypt

**Reply by daastek** [reply]
If you look at the jaeger span, bcrypt took only 1.18ms and we are using bcrypt cost 4. 
Also the bcrypt performance should be same for both cockroach and mysql. isn't it ?

**Reply by aeneasr** [reply]
Please include more details on your issue, at best a reproducible repository. If it’s a cockroach slow query issue it also makes sense to check with them instead

**Reply by daastek** [reply]
Hydra version: 1.10.1 (cpu 1, memory 200MB)
Cockroachdb version:  20.2.7 (cpu 3, memory, 12GB)

I started a new auto scaling hydra deployment with bcrypt cost 4 and cocckroachdb 3 node secure cluster in same kubernets cluster.  DSN used in hydra is `cockroach://root:<password>@cockroachdb-public:26257/defaultdb?sslmode=require&sslcert=/cockroach/cockroach-certs/client.root.crt&sslkey=/cockroach/cockroach-certs/client.root.key&sslrootcert=/cockroach/cockroach-certs/ca.crt` 

Then ran a performance test with 150 client_credentials requests per second for 5 mins.

The result was around 4 second response time.

There is no indication of slow queries in cockroachdb. The jaeger spans looks like there is some waits between the spans. I checked the kubernetes cpu throttling, there was none. Hydra scaled to 5 pods during this 5 mins. 

Please let me know if you need any more information or I have need to tweak some settings and run the tests again.

**Reply by aeneasr** [reply]
Maybe check the transactions in crdb dashboard?

**Reply by daastek** [reply]
The longest transaction in crdb dashboard is only 180ms.

**Reply by aeneasr** [reply]
Hm, interesting. Could you run CPU profiling? So setting env var „PROFILING=cpu“ and catching and rendering the output: https://blog.golang.org/pprof

That will help identify CPU sinks - but only makes sense if the Hydra process is taking lots of CPU. If it isn’t it’s more likely to be a locking issue

**Reply by aeneasr** [reply]
Or maybe it’s some transaction locking so that some transactions need to wait for others to finish? This could be something CRDB specific!

**Reply by aeneasr** [reply]
Also please share how you run the test!

**Reply by daastek** [reply]
We use [Predator](https://www.predator.dev/) for load testing. I will try the CPU profiling and get back.

**Reply by vinckr** [reply]
Hey @daastek,
 just checking in if your problem still persists. 
I will mark this as answered for now, but feel free to come back to the discussion any time.
Thanks a ton :bee:
